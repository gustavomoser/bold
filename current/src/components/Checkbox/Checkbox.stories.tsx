import { action } from '@storybook/addon-actions'
import { boolean, text } from '@storybook/addon-knobs'
import React from 'react'

import { Checkbox } from './Checkbox'

export default {
  title: 'Components/Checkbox',
}

export const Default = () => (
  <Checkbox
    name='check'
    label={text('label', 'Component label')}
    disabled={boolean('disabled', false)}
    indeterminate={boolean('indeterminate', false)}
    onChange={action('changed')}
  />
)
