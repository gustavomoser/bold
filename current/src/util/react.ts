import { MutableRefObject, Ref } from 'react'

/**
 * Compose multiple react refs.
 *
 * @returns A new ref that passes down the Element to all ref arguments.
 */
export function composeRefs<T>(...refs: Array<Ref<T>>) {
  return (value: T) => refs.forEach(ref => setRef(ref, value))
}

/**
 * Changes the value of a ref.
 * The ref can be a function ref or a object ref.
 *
 * @param ref The ref to have its value changed.
 * @param value The new value.
 */
export function setRef<T>(ref: Ref<T>, value: any) {
  if (typeof ref === 'function') {
    ref(value)
  } else if (ref) {
    const mutableRef = ref as MutableRefObject<T>
    mutableRef.current = value
  }
}

/**
 * Compose multiple event handlers.
 *
 * @returns A new function that triggers all handlers passed as parameters, in order.
 */
export function composeHandlers(...handlers: Array<(...params: any) => void>) {
  return (...args: any[]) => handlers.forEach(handler => handler && handler(...args))
}

/**
 * Invokes the native `value` property setter of an element.
 * From https://github.com/facebook/react/issues/10135#issuecomment-401496776
 *
 * @param element The element to invoke the setter on.
 * @param value The value to be set.
 */
export function setNativeValue(element: HTMLElement, value: any) {
  const prototype = Object.getPrototypeOf(element)
  const { set: prototypeValueSetter = null } = Object.getOwnPropertyDescriptor(prototype, 'value') || {}
  const { set: valueSetter = null } = Object.getOwnPropertyDescriptor(element, 'value') || {}

  if (prototypeValueSetter && valueSetter !== prototypeValueSetter) {
    prototypeValueSetter.call(element, value)
  } else if (valueSetter) {
    valueSetter.call(element, value)
  } else {
    throw new Error('The given element does not have a value setter')
  }
}
