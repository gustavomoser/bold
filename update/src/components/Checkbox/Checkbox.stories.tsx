import { action } from '@storybook/addon-actions'
import { Meta, StoryObj } from '@storybook/react'
import { Checkbox, CheckboxProps } from './Checkbox'

const meta: Meta<CheckboxProps> = {
  title: 'Components/Checkbox',
  component: Checkbox,
  tags: ['autodocs'],
  args: {
    name: 'check',
    label: 'Component label',
    disabled: false,
    indeterminate: false,
    onChange: action('changed'),
  },
}

export default meta
type Story = StoryObj<CheckboxProps>

export const Unchecked: Story = {}
export const Checked: Story = {
  args: {
    checked: true,
  },
}
export const Indeterminate: Story = {
  args: {
    indeterminate: true,
  },
}
export const Disabled: Story = {
  args: {
    disabled: true,
  },
}
