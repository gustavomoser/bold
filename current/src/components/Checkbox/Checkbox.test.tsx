import { render, fireEvent } from '@testing-library/react'
import React from 'react'

import { Checkbox } from './Checkbox'

describe('Checkbox', () => {
  it('should render correctly', () => {
    const { container } = render(<Checkbox label='check' />)
    expect(container).toMatchSnapshot()
  })
  it('should render correctly when disabled', () => {
    const { container } = render(<Checkbox label='check' disabled />)
    expect(container).toMatchSnapshot()
  })
  it('should render correctly when indeterminate', () => {
    const { container } = render(<Checkbox label='check' indeterminate />)
    expect(container).toMatchSnapshot()
  })
  it('should have the indeterminate attribute on input when prop is specified', () => {
    const { container } = render(<Checkbox label='check' indeterminate />)
    const input = container.querySelector('input')
    expect(input.indeterminate).toEqual(true)
  })
  it('should use external inputRef', () => {
    const ref = React.createRef<HTMLInputElement>()

    const { container } = render(<Checkbox label='check' inputRef={ref} checked={false} />)
    const input = container.querySelector('input')
    fireEvent(input, new MouseEvent('click'))
    expect(input.checked).toBeTruthy()
    expect(ref.current.checked).toBeTruthy()
  })
})
