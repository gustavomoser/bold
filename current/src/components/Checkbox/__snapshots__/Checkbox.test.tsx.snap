// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should render correctly 1`] = `
.emotion-3 {
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-1 {
  background-color: #FFFFFF;
  border: 1px solid #8F8FA2;
  border-radius: 2px;
  display: inline-block;
  height: 24px;
  position: relative;
  -webkit-transition: all .2s ease;
  transition: all .2s ease;
  vertical-align: middle;
  width: 24px;
  min-width: 24px;
}

.emotion-2 {
  margin-left: 0.5rem;
}

.emotion-0 {
  opacity: 0;
  margin-right: -13px;
}

.emotion-0:hover + .emotion-1 {
  border-color: #7A7B8E;
}

.emotion-0:checked + .emotion-1 {
  background-color: #0069D0;
  border-color: #0069D0;
}

.emotion-0:checked + .emotion-1:after {
  content: "";
  border-right: 2px solid #FFFFFF;
  border-bottom: 2px solid #FFFFFF;
  position: absolute;
  width: 8px;
  height: 16px;
  top: 0;
  left: 7px;
  opacity: 1;
  -webkit-transform: rotate(45deg) scale(1);
  -ms-transform: rotate(45deg) scale(1);
  transform: rotate(45deg) scale(1);
  -webkit-transition: opacity .2s ease;
  transition: opacity .2s ease;
}

.emotion-0:indeterminate + .emotion-1 {
  background-color: #FFFFFF;
  border-color: #0069D0;
}

.emotion-0:indeterminate + .emotion-1:after {
  content: "";
  border-bottom: 3px solid #0069D0;
  position: absolute;
  top: -1px;
  left: 4px;
  width: 14px;
  height: 14px;
  opacity: 1;
  -webkit-transition: opacity .2s ease;
  transition: opacity .2s ease;
}

.emotion-0:focus + .emotion-1 {
  box-shadow: 0 0 0 2px #F0F0F5,0 0 0 4px #0069D0;
}

.emotion-0:disabled + .emotion-1 {
  opacity: 0.4;
  cursor: not-allowed;
}

.emotion-0:disabled + .emotion-1 + .emotion-2 {
  color: #AAAAB9;
  cursor: not-allowed;
}

<div>
  <label
    class="emotion-3"
  >
    <input
      class="emotion-0"
      type="checkbox"
    />
    <span
      class="emotion-1"
    />
    <span
      class="emotion-2"
    >
      check
    </span>
  </label>
</div>
`;

exports[`Checkbox should render correctly when disabled 1`] = `
.emotion-1 {
  background-color: #FFFFFF;
  border: 1px solid #8F8FA2;
  border-radius: 2px;
  display: inline-block;
  height: 24px;
  position: relative;
  -webkit-transition: all .2s ease;
  transition: all .2s ease;
  vertical-align: middle;
  width: 24px;
  min-width: 24px;
}

.emotion-2 {
  margin-left: 0.5rem;
}

.emotion-0 {
  opacity: 0;
  margin-right: -13px;
}

.emotion-0:hover + .emotion-1 {
  border-color: #7A7B8E;
}

.emotion-0:checked + .emotion-1 {
  background-color: #0069D0;
  border-color: #0069D0;
}

.emotion-0:checked + .emotion-1:after {
  content: "";
  border-right: 2px solid #FFFFFF;
  border-bottom: 2px solid #FFFFFF;
  position: absolute;
  width: 8px;
  height: 16px;
  top: 0;
  left: 7px;
  opacity: 1;
  -webkit-transform: rotate(45deg) scale(1);
  -ms-transform: rotate(45deg) scale(1);
  transform: rotate(45deg) scale(1);
  -webkit-transition: opacity .2s ease;
  transition: opacity .2s ease;
}

.emotion-0:indeterminate + .emotion-1 {
  background-color: #FFFFFF;
  border-color: #0069D0;
}

.emotion-0:indeterminate + .emotion-1:after {
  content: "";
  border-bottom: 3px solid #0069D0;
  position: absolute;
  top: -1px;
  left: 4px;
  width: 14px;
  height: 14px;
  opacity: 1;
  -webkit-transition: opacity .2s ease;
  transition: opacity .2s ease;
}

.emotion-0:focus + .emotion-1 {
  box-shadow: 0 0 0 2px #F0F0F5,0 0 0 4px #0069D0;
}

.emotion-0:disabled + .emotion-1 {
  opacity: 0.4;
  cursor: not-allowed;
}

.emotion-0:disabled + .emotion-1 + .emotion-2 {
  color: #AAAAB9;
  cursor: not-allowed;
}

.emotion-3 {
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: not-allowed;
}

<div>
  <label
    class="emotion-3"
  >
    <input
      class="emotion-0"
      disabled=""
      type="checkbox"
    />
    <span
      class="emotion-1"
    />
    <span
      class="emotion-2"
    >
      check
    </span>
  </label>
</div>
`;

exports[`Checkbox should render correctly when indeterminate 1`] = `
.emotion-3 {
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-1 {
  background-color: #FFFFFF;
  border: 1px solid #8F8FA2;
  border-radius: 2px;
  display: inline-block;
  height: 24px;
  position: relative;
  -webkit-transition: all .2s ease;
  transition: all .2s ease;
  vertical-align: middle;
  width: 24px;
  min-width: 24px;
}

.emotion-2 {
  margin-left: 0.5rem;
}

.emotion-0 {
  opacity: 0;
  margin-right: -13px;
}

.emotion-0:hover + .emotion-1 {
  border-color: #7A7B8E;
}

.emotion-0:checked + .emotion-1 {
  background-color: #0069D0;
  border-color: #0069D0;
}

.emotion-0:checked + .emotion-1:after {
  content: "";
  border-right: 2px solid #FFFFFF;
  border-bottom: 2px solid #FFFFFF;
  position: absolute;
  width: 8px;
  height: 16px;
  top: 0;
  left: 7px;
  opacity: 1;
  -webkit-transform: rotate(45deg) scale(1);
  -ms-transform: rotate(45deg) scale(1);
  transform: rotate(45deg) scale(1);
  -webkit-transition: opacity .2s ease;
  transition: opacity .2s ease;
}

.emotion-0:indeterminate + .emotion-1 {
  background-color: #FFFFFF;
  border-color: #0069D0;
}

.emotion-0:indeterminate + .emotion-1:after {
  content: "";
  border-bottom: 3px solid #0069D0;
  position: absolute;
  top: -1px;
  left: 4px;
  width: 14px;
  height: 14px;
  opacity: 1;
  -webkit-transition: opacity .2s ease;
  transition: opacity .2s ease;
}

.emotion-0:focus + .emotion-1 {
  box-shadow: 0 0 0 2px #F0F0F5,0 0 0 4px #0069D0;
}

.emotion-0:disabled + .emotion-1 {
  opacity: 0.4;
  cursor: not-allowed;
}

.emotion-0:disabled + .emotion-1 + .emotion-2 {
  color: #AAAAB9;
  cursor: not-allowed;
}

<div>
  <label
    class="emotion-3"
  >
    <input
      class="emotion-0"
      type="checkbox"
    />
    <span
      class="emotion-1"
    />
    <span
      class="emotion-2"
    >
      check
    </span>
  </label>
</div>
`;
